arelle/plugin/validate/HMRC/ValidateHmrc.py

```diff
@@ -314,9 +330,96 @@ def _evaluateCode(self, code: str) -> CodeResult:
+        if (not (self._getAndCheckValidFacts(CONCEPT_NAME_ENTITY_AUDITORS) and self._getAndCheckValidFacts(CONCEPT_NAME_SENIOR_STATUTORY_AUDITOR))
+                and not self._getAndCheckValidFacts(CONCEPT_NAME_INDIVIDUAL_AUDITOR)):
+            missingConcepts.append(CONCEPT_NAME_INDIVIDUAL_AUDITOR)
+            missingConcepts.append(CONCEPT_NAME_SENIOR_STATUTORY_AUDITOR)
+            missingConcepts.append(CONCEPT_NAME_ENTITY_AUDITORS)
+        if len(missingConcepts) > 0:
+            return self._setCode(
+                code,
+                CodeResult(
+                    success=False,
+                    conceptList=missingConcepts,
+                    message="An audited report must contain facts tagged with the concepts of DateAuditorsReport, OpinionAuditorsOnEntity"
+                            "as well as either NameIndividualAuditor or the combination of NameSeniorStatutoryAuditor and NameEntityAuditors"
+                            "There are no facts tagged with the concepts: %(conceptLocalName)s"
+                )
+            )
+        return self._setCode(code, CodeResult())
+
+    def _evaluateMedFact(self, prefix: str) -> CodeResult:
+        """
+        Logs a warning when an medium company or LLP does not facts tagged with the concept of
+        "StatementThatCompanyHasPreparedAccountsUnderProvisionsRelatingToMedium-sizedCompanies".
+        :param prefix:
+        """
+        code = ''
+        if prefix == COMPANY:
+            code = CO_MED_CO
+        elif prefix == LLP:
+            code = LP_MED_LP
+        if not self._getAndCheckValidFacts(CONCEPT_MEDIUM_COMPANY):
+            return self._setCode(
+                code,
+                CodeResult(
+                    success=False,
+                    warning=True,
+                    message="The concept of StatementThatCompanyHasPreparedAccountsUnderProvisionsRelatingToMedium-sizedCompanies must exist and have a non-nil value."
+                )
+            )
+        return self._setCode(code, CodeResult())
+
+    def _evaluateCode(self, code: str) -> CodeResult:
+        """
+        Evaluates whether the conditions associated with the given code pass.
+        Result is cached.
+        :param code:
+        :return: Result object that can be used for additional logic or to trigger an error.
+        """
+        if code in self._codeResultMap:
+            return self._codeResultMap[code]
+        textValidationPatterns = TEXT_VALIDATION_PATTERNS.get(code, {})
+        if textValidationPatterns:
+            result = self._evaluateTextPattern(code, textValidationPatterns)
+        if code == CO_AUDIT:
+            result = self._evaluateAuditFacts(COMPANY)
+        if code == CO_MED_CO:
+            result = self._evaluateMedFact(COMPANY)
+        if code == LP_AUDIT:
+            result = self._evaluateAuditFacts(LLP)
+        if code == LP_MED_LP:
+            result = self._evaluateMedFact(LLP)
+        return result
```

Comment:
Instead of calling `_setCode` in the evaluate sub-functions, I think it's better to have it in only one place at the end of this function. Along with this, the above chain of `if` statements should be `elif` since we would not expect one code to need multiple of these evaluations

```suggestion
        return self._setCode(code, result)
```